### PR TITLE
[ADI] Add vote creation and suppression endpoints

### PR DESCRIPTION
**Pull request by illusionOfParadise**

### CODE REVIEW

File: src/Normalizer/AdherentNormalizer.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    public function denormalize($data, $class, $format = null, array $context = [])
+    {
+        if (!isset($data['id'])) {
+            $data['adherent'] = $this->tokenStorage->getToken()->getUser()->getId();
+        }
+
+        $data = $this->normalizer->denormalize($data, $class, $format, $context);
+//        dump($data);die;
```
Comment (thomasbisignani) [inline src/Normalizer/AdherentNormalizer.php]: :man_cartwheeling:

File: src/Entity/IdeasWorkshop/Vote.php [hunk 2]
```diff
@@ -67,6 +94,7 @@ public function getIdea(): Idea
     public function setIdea(Idea $idea): void
     {
         $this->idea = $idea;
+        $this->idea->incrementVotesCount(1);
```
Comment (thomasbisignani) [inline src/Entity/IdeasWorkshop/Vote.php]: This default argument value is useless here

 ```php
public function incrementVotesCount(int $increment = 1): void
```

File: features/api/ideas_workshop_needs.feature [hunk 3]
```diff
@@ -1,14 +1,14 @@
 @ap
```
Comment (thomasbisignani) [inline features/api/ideas_workshop_needs.feature]: Not a diff in your PR, but can you fix this line by `@api` in your PR please ?

File: app/config/services/services.xml [hunk 4]
```diff
@@ -212,6 +212,12 @@
         <service id="AppBundle\Serializer\IdeasWorkshopSerializerNormalizer">
             <argument type="service" id="api_platform.serializer.normalizer.item" />
         </service>
+        <service id="AppBundle\Normalizer\AdherentNormalizer">
+            <argument type="service" id="api_platform.serializer.normalizer.item"/>
+            <argument type="service" id="security.token_storage" />
```
Comment (ottaviano) [inline app/config/services/services.xml]: you can omit the second argument and tag. the argument will be injected by autowiring and the tag will be added by autoconfiguring
```
<argument type="service" id="security.token_storage" />
<tag name="serializer.normalizer"/>
```

File: src/Entity/Adherent.php [hunk 5]
```diff
@@ -30,7 +30,14 @@
 use JMS\Serializer\Annotation as JMS;
 
 /**
- * @ApiResource
+ * @ApiResource(
+ *     collectionOperations={},
+ *     itemOperations={
+ *         "get": {
```
Comment (ottaviano) [inline src/Entity/Adherent.php]: do we really need to have this endpoint? for what?

File: src/Entity/IdeasWorkshop/Vote.php [hunk 6]
```diff
@@ -7,10 +7,28 @@
...
+ *         "normalization_context": {
+ *             "groups": {"vote_read"}
+ *         },
+ *         "denormalization_context": {
+ *             "groups": {"vote_write"}
+ *         },
+ *     },
+ *     collectionOperations={
+ *         "get",
+ *         "post": {"access_control": "is_granted('IS_AUTHENTICATED_FULLY')"}
```
Comment (ottaviano) [inline src/Entity/IdeasWorkshop/Vote.php]: what about `is_granted({'ROLE_ADHERENT', 'ROLE_ADMIN_DASHBOARD'})`?
as `IS_AUTHENTICATED_FULLY` can be a simple user.

File: src/Entity/IdeasWorkshop/Vote.php [hunk 7]
```diff
@@ -32,25 +50,34 @@ class Vote
     /**
      * @ORM\ManyToOne(targetEntity="Idea", inversedBy="votes")
      * @ORM\JoinColumn(nullable=false)
+     *
+     * @Assert\NotNull
```
Comment (ottaviano) [inline src/Entity/IdeasWorkshop/Vote.php]: maybe `NotBlanc`? `NotNull` allows to have a empty string `""`

File: src/Entity/IdeasWorkshop/Vote.php [hunk 8]
```diff
@@ -32,25 +50,34 @@ class Vote
...
+     *
+     * @SymfonySerializer\Groups({"idea_list_read", "vote_read", "vote_write"})
      */
     private $type;
 
-    public function __construct(Adherent $adherent, string $type)
+    public function __construct(string $type = null)
     {
-        $this->adherent = $adherent;
         $this->type = $type;
```
Comment (ottaviano) [inline src/Entity/IdeasWorkshop/Vote.php]: what about checking the type value with `VoteTypeEnum::isValid($type)`?

File: src/Entity/IdeasWorkshop/Vote.php (hunk 2 above)
Comment (ottaviano) [reply src/Entity/IdeasWorkshop/Vote.php]: can you move this call to Idea class pls? in `addVote` method

File: src/Normalizer/AdherentNormalizer.php [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+
+final class AdherentNormalizer implements DenormalizerInterface
+{
+    private $normalizer;
+    private $tokenStorage;
+
+    public function __construct(NormalizerInterface $normalizer, TokenStorageInterface $tokenStorage)
+    {
+        if (!$normalizer instanceof DenormalizerInterface) {
```
Comment (ottaviano) [inline src/Normalizer/AdherentNormalizer.php]: why do you not typehint with this interface?

File: app/config/services/services.xml (hunk 4 above)
Comment (illusionOfParadise) [reply app/config/services/services.xml]: Done.

File: src/Entity/IdeasWorkshop/Vote.php (hunk 7 above)
Comment (illusionOfParadise) [reply src/Entity/IdeasWorkshop/Vote.php]: This field in the database is integer, an empty string could not be safe and it's weird to have `NotBlank` on a non string property, no?

File: src/Entity/Adherent.php (hunk 5 above)
Comment (illusionOfParadise) [reply src/Entity/Adherent.php]: Without this config the error `No item route associated with the type \AppBundle\Entity\Adherent` occurs while posting a vote. It's impossible to create a response when the post was successful.

File: src/Normalizer/AdherentNormalizer.php (hunk 9 above)
Comment (illusionOfParadise) [reply src/Normalizer/AdherentNormalizer.php]: Done.

File: src/Entity/IdeasWorkshop/Vote.php (hunk 6 above)
Comment (illusionOfParadise) [reply src/Entity/IdeasWorkshop/Vote.php]: Done.

File: src/Entity/IdeasWorkshop/Vote.php (hunk 2 above)
Comment (illusionOfParadise) [reply src/Entity/IdeasWorkshop/Vote.php]: Changed. Now a listener does incrementation.

File: src/Entity/IdeasWorkshop/Vote.php (hunk 8 above)
Comment (illusionOfParadise) [reply src/Entity/IdeasWorkshop/Vote.php]: Now, it's checked with validation while setting a value https://github.com/EnMarche/en-marche.fr/pull/2946/files#diff-39c4e64c8fc7a47c56a13eaa32ae313aL47

File: features/api/ideas_workshop_needs.feature [hunk 10]
```diff
@@ -1,14 +1,14 @@
-@ap
+@app
```
Comment (thomasbisignani) [inline features/api/ideas_workshop_needs.feature]: `@api` no ?

IMO it is fixed by https://github.com/EnMarche/en-marche.fr/pull/2945/

File: src/EntityListener/VoteListener.php [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+namespace AppBundle\EntityListener;
+
+use AppBundle\Entity\IdeasWorkshop\Vote;
+use Doctrine\Common\Persistence\Event\LifecycleEventArgs;
+
+class VoteListener
+{
+    public function prePersist(Vote $vote, LifecycleEventArgs $args)
+    {
+        $vote->getIdea()->incrementVotesCount(1);
```
Comment (thomasbisignani) [inline src/EntityListener/VoteListener.php]: The default parameter value is not needed

File: src/EntityListener/VoteListener.php [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+    }
+
+    public function preRemove(Vote $vote, LifecycleEventArgs $args)
+    {
+        $vote->getIdea()->decrementVotesCount(1);
```
Comment (thomasbisignani) [inline src/EntityListener/VoteListener.php]: Same here

Comment (illusionOfParadise) [reply src/EntityListener/VoteListener.php]: Done.

File: src/EntityListener/VoteListener.php (hunk 11 above)
Comment (illusionOfParadise) [reply src/EntityListener/VoteListener.php]: Done.

**Pull request merged**